packages/material-ui-lab/src/TreeView/TreeView.js

```diff
@@ -394,164 +523,218 @@ const TreeView = React.forwardRef(function TreeView(props, ref) {
+        event.stopPropagation();
+        break;
+      case 'Enter':
+        if (nodeMap[focusedNodeId].expandable) {
+          toggleExpansion(event);
+          flag = true;
         }
-      }
-      return list;
-    };
+        event.stopPropagation();
+        break;
+      case 'ArrowDown':
+        if (multiSelect && event.shiftKey && !disableSelection) {
+          selectNextNode(event, focusedNodeId);
+        }
+        focusNextNode(event, focusedNodeId);
+        flag = true;
+        break;
+      case 'ArrowUp':
+        if (multiSelect && event.shiftKey && !disableSelection) {
+          selectPreviousNode(event, focusedNodeId);
+        }
+        focusPreviousNode(event, focusedNodeId);
+        flag = true;
+        break;
+      case 'ArrowRight':
+        if (theme.direction === 'rtl') {
+          flag = handlePreviousArrow(event);
+        } else {
+          flag = handleNextArrow(event);
+        }
+        break;
+      case 'ArrowLeft':
+        if (theme.direction === 'rtl') {
+          flag = handleNextArrow(event);
+        } else {
+          flag = handlePreviousArrow(event);
+        }
+        break;
+      case 'Home':
+        if (multiSelect && ctrlPressed && event.shiftKey && !disableSelection) {
+          rangeSelectToFirst(event, focusedNodeId);
+        }
+        focusFirstNode(event);
+        flag = true;
+        break;
+      case 'End':
+        if (multiSelect && ctrlPressed && event.shiftKey && !disableSelection) {
+          rangeSelectToLast(event, focusedNodeId);
+        }
+        focusLastNode(event);
+        flag = true;
+        break;
+      default:
+        if (key === '*') {
+          expandAllSiblings(event, focusedNodeId);
+          flag = true;
+        } else if (multiSelect && ctrlPressed && key.toLowerCase() === 'a' && !disableSelection) {
+          flag = selectAllNodes(event);
+        } else if (!ctrlPressed && !event.shiftKey && isPrintableCharacter(key)) {
+          focusByFirstCharacter(event, focusedNodeId, key);
+          flag = true;
+        }
+    }
 
-    if (childrenCalculated) {
-      visibleNodes.current = buildVisible(nodeMap.current[-1].children);
+    if (flag) {
+      event.preventDefault();
+      event.stopPropagation();
     }
-  }, [expanded, childrenCalculated, isExpanded, children]);
 
-  const noopSelection = () => {
-    return false;
+    if (onKeyDown) {
+      onKeyDown(event);
+    }
+  };
+
+  const handleFocus = (event) => {
+    const firstSelected = Array.isArray(selected) ? selected[0] : selected;
+    focus(event, firstSelected || getChildrenIds(null)[0]);
+
+    if (onFocus) {
+      onFocus(event);
+    }
+  };
+
+  const handleBlur = (event) => {
+    setFocusedNodeId(null);
+
+    if (onBlur) {
+      onBlur(event);
+    }
   };
 
   return (
     <TreeViewContext.Provider
       value={{
         icons: { defaultCollapseIcon, defaultExpandIcon, defaultParentIcon, defaultEndIcon },
         focus,
-        focusFirstNode,
```

Comment:
A lot less stuff being passed through context is great for readability and potentially performance :)